larpandoracontent/LArHelpers/LArClusterHelper.cc

```diff
@@ -642,6 +643,27 @@ void LArClusterHelper::GetCaloHitListInBoundingBox(const pandora::Cluster *const
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+std::set<unsigned int> LArClusterHelper::GetDaughterVolumeIDs(const Cluster *const pCluster)
+{
+    std::set<unsigned int> daughterVolumeIds;
+    const OrderedCaloHitList &orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+
+    for (OrderedCaloHitList::const_iterator ochIter = orderedCaloHitList.begin(), ochIterEnd = orderedCaloHitList.end(); ochIter != ochIterEnd; ++ochIter)
+    {
+        for (CaloHitList::const_iterator hIter = ochIter->second->begin(), hIterEnd = ochIter->second->end(); hIter != hIterEnd; ++hIter)
+        {
+            const CaloHit *const pCaloHit = *hIter;
+            const LArCaloHit *const pLArCaloHit(dynamic_cast<const LArCaloHit*>(pCaloHit));
+
+            if (pLArCaloHit) 
+                daughterVolumeIds.insert(pLArCaloHit->GetDaughterVolumeId());
```

Comment:
If the dynamic cast fails, I guess this implies a significant issue with the hits that probably warrants an exception?

Reply:
Etienne will comment here, but it looks like checks of this type are performed in MasterAlgorithm.cc as well.
As a separate note, I was wondering if we care to avoid storing the same DaughterVolumeId more than once?

Reply:
The check should definitely be performed, but in the MasterAlgorithm, if this check fails (aside from one specific case), exceptions are thrown, because it usually implies a fundamental problem if you weren't able to cast the CaloHit to a LArCaloHit, and so shouldn't attempt further processing.

Reply:
Yep, I can see an exception is thrown in the copy constructor in the Master Algorithm, and I think it's definitely necessary there, perhaps less so here - but of course happy for us to add it if that's the consensus!

Reply:
Yep, I can see an exception is thrown in the copy constructor in the Master Algorithm, and I think it's definitely necessary there, perhaps less so here - but of course happy for us to add it if that's the consensus!